fvm/evm/handler/blockHashList.go

```diff
@@ -0,0 +1,266 @@
+	}
+	// updates the block hash stored at index
+	err := bhl.updateBlockHashAt(bhl.tail, bh)
+	if err != nil {
+		return err
+	}
+
+	// update meta data
+	bhl.tail = (bhl.tail + 1) % bhl.capacity
+	bhl.height = height
+	if bhl.count != bhl.capacity {
+		bhl.count++
+	}
+	return bhl.storeMetaData()
+}
+
+// IsEmpty returns true if the list is empty
+func (bhl *BlockHashList) IsEmpty() bool {
+	return bhl.count == 0
+}
+
+// LastAddedBlockHash returns the last block hash added to the list
+// for empty list it returns empty hash value
+func (bhl *BlockHashList) LastAddedBlockHash() (gethCommon.Hash, error) {
+	if bhl.count == 0 {
+		// return empty hash
+		return gethCommon.Hash{}, nil
+	}
+	index := bhl.tail - 1
+	if index < 0 {
+		index = bhl.capacity - 1
+	}
+	return bhl.getBlockHashAt(index)
+}
+
+// MinAvailableHeight returns the min available height in the list
+func (bhl *BlockHashList) MinAvailableHeight() uint64 {
+	if bhl.IsEmpty() {
+		return 0
+	}
+	return bhl.height - (uint64(bhl.count) - 1)
+}
+
+// MaxAvailableHeight returns the max available height in the list
+func (bhl *BlockHashList) MaxAvailableHeight() uint64 {
+	return bhl.height
+}
+
+// BlockHashByIndex returns the block hash by block height
+func (bhl *BlockHashList) BlockHashByHeight(height uint64) (found bool, bh gethCommon.Hash, err error) {
+	if bhl.count == 0 || // if empty
+		height > bhl.MaxAvailableHeight() || // or height too high
+		height < bhl.MinAvailableHeight() { // or height too low
+		return false, gethCommon.Hash{}, nil
+	}
+	// calculate the index to lookup
+	diff := bhl.height - height
+	index := bhl.tail - int(diff) - 1
+	if index < 0 {
+		index = bhl.capacity + index
+	}
+	bh, err = bhl.getBlockHashAt(index)
+	return true, bh, err
+}
+
+// updateBlockHashAt updates block hash at index
+func (bhl *BlockHashList) updateBlockHashAt(idx int, bh gethCommon.Hash) error {
+	// fetch the bucket
+	bucketNumber := idx / hashCountPerBucket
+	bucket, err := bhl.fetchBucket(bucketNumber)
+	if err != nil {
+		return err
+	}
+	// update the block hash
+	start := (idx % hashCountPerBucket) * hashEncodingSize
+	end := start + hashEncodingSize
+	copy(bucket[start:end], bh.Bytes())
+
+	// store bucket
+	return bhl.backend.SetValue(
+		bhl.rootAddress[:],
+		[]byte(fmt.Sprintf(blockHashListBucketKeyFormat, bucketNumber)),
+		bucket,
+	)
+}
+
+// fetchBucket fetches the bucket
+func (bhl *BlockHashList) fetchBucket(num int) ([]byte, error) {
+	data, err := bhl.backend.GetValue(
+		bhl.rootAddress[:],
+		[]byte(fmt.Sprintf(blockHashListBucketKeyFormat, num)),
+	)
+	if err != nil {
+		return nil, err
+	}
+	// if not exist create and return an new empty buffer
+	if len(data) == 0 {
+		return make([]byte, hashCountPerBucket*hashEncodingSize), nil
+	}
+	return data, err
+}
+
+// returns the block hash at the given index
+func (bhl *BlockHashList) getBlockHashAt(idx int) (gethCommon.Hash, error) {
+	// fetch the bucket first
```

Comment:
should we check if the index is within the bounds

Reply:
I think we can skip that, this is internal method and checks on the higher-level methods ensure its always in the expected range